dot/parachain/provisioner/prioritized_disputes_selection.go

```diff
@@ -15,10 +17,236 @@ import (
+				func(validatorIdx parachaintypes.ValidatorIndex, vote parachaintypes.Vote[parachaintypes.ValidDisputeStatementKind]) bool {
+					validDisputeStatement := &parachaintypes.DisputeStatement{}
+					err := validDisputeStatement.SetValue(parachaintypes.ValidDisputeStatement{
+						Kind: vote.Kind,
+					})
+					if err != nil {
+						panic(
+							fmt.Sprintf(
+								"%T is an invalid variant of %T",
+								vote.Kind,
+								validDisputeStatement,
+							),
+						)
+					}
+
+					isVoteWorth := isVoteWorthToKeep(
+						validatorIdx,
+						*validDisputeStatement,
+						onchainState,
+					)
+					if !isVoteWorth {
+						validatorIdxToRemove = append(validatorIdxToRemove, validatorIdx)
+					}
+					return true
+				},
+			)
+
+			for _, vi := range validatorIdxToRemove {
+				votes.Valid.Delete(vi)
+			}
+
+			validatorIdxToRemove = make([]parachaintypes.ValidatorIndex, 0)
+			votes.Invalid.Ascend(
+				parachaintypes.ValidatorIndex(0),
+				func(validatorIdx parachaintypes.ValidatorIndex, vote parachaintypes.Vote[parachaintypes.InvalidDisputeStatementKind]) bool {
+					invalidDisputeStatement := &parachaintypes.DisputeStatement{}
+					err := invalidDisputeStatement.SetValue(parachaintypes.InvalidDisputeStatement{
+						Kind: vote.Kind,
+					})
+					if err != nil {
+						panic(
+							fmt.Sprintf(
+								"%T is an valid variant of %T",
+								vote.Kind,
+								invalidDisputeStatement,
+							),
+						)
+					}
+
+					if !isVoteWorthToKeep(validatorIdx, *invalidDisputeStatement, onchainState) {
+						validatorIdxToRemove = append(validatorIdxToRemove, validatorIdx)
+						// votes.Valid.Delete(validatorIdx)
+					}
+					return true
+				},
+			)
+
+			for _, vi := range validatorIdxToRemove {
+				votes.Invalid.Delete(vi)
+			}
+
+			selectedVotes = append(selectedVotes,
+				voteSelectionResult{
+					key: parachaintypes.DisputeKey{
+						SessionIndex:  sessionIndex,
+						CandidateHash: candidateHash,
+					},
+					votes: votes,
+				})
+		}
+
+		// Check if votes are within the limit
+		for _, vote := range selectedVotes {
+			sessionIndex := vote.key.SessionIndex
+			candidateHash := vote.key.CandidateHash
+			selectedVotes := vote.votes
+
+			votesLen := selectedVotes.Valid.Len() + selectedVotes.Invalid.Len()
+			if votesLen+totalVotesLen > maxDisputesVotes {
+				// we are done - no more votes can be added. Importantly, we don't add any votes for
+				// a dispute here if we can't fit them all. This gives us an important invariant,
+				// that backing votes for disputes make it into the provisioned vote set.
+
+				return sortVoteSelectionResults(result), nil
+			}
+
+			result[parachaintypes.DisputeKey{SessionIndex: sessionIndex, CandidateHash: candidateHash}] = selectedVotes
+			totalVotesLen += votesLen
+		}
+	}
+
+	return sortVoteSelectionResults(result), nil
+}
+
+// sortVoteSelectionResults sorts the vote selection results based on SessionIndex and CandidateHash.
```

Comment:
could we not just use `btree.Map`?

Reply:
Could you not just a comparable version of `DisputeKey`? I wonder which would be more performant.

Reply:
I think that is not possible given the current btree.Map implementation, they contraints the map key to be that `orderable` type (`~int`, `~uint`, ..., `~string`) which is possible for types like `type A string` or `type B uint` but not for types like `type A struct { Value string }`.